File: api/nb-weather-ol/src/main/java/com/example/nb_weather_ol/service/WeatherQueriesService.java
package com.example.nb_weather_ol.service;

import java.util.List;
import java.util.Optional;

import org.slf4j.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.nb_weather_ol.model.*;
import com.example.nb_weather_ol.repository.*;

@Service
public class WeatherQueriesService {
    private static final Logger logger = LoggerFactory.getLogger(WeatherQueriesService.class);

    private final WeatherQueriesRepository weatherQueriesRepository;

    @Autowired
    public WeatherQueriesService(WeatherQueriesRepository weatherQueriesRepository) {
        this.weatherQueriesRepository = weatherQueriesRepository;
    }

    // Create
    public WeatherQueries createWeatherQuery(WeatherQueriesRequest request) {
        logger.info("Creating or retrieving weather query entry");
        try {
            Optional<WeatherQueries> existingQuery = weatherQueriesRepository
                    .findByQueryAndUserId(request.getQuery(), request.getUserId());

            if (existingQuery.isPresent()) {
                logger.info("Existing weather query found for query: {} and userId: {}",
                        request.getQuery(), request.getUserId());
                return existingQuery.get();
            } else {
                logger.info("Creating new weather query entry for query: {} and userId: {}",
                        request.getQuery(), request.getUserId());
                return weatherQueriesRepository.save(request.getQuery(), request.getUserId());
            }
        } catch (Exception e) {
            logger.error("Error creating or retrieving weather query entry: {}", e.getMessage());
            throw new RuntimeException("Failed to create or retrieve weather query entry", e);
        }
    }

    // Read (single entry)
    public Optional<WeatherQueries> getWeatherQueryById(Long id) {
        logger.info("Fetching weather query with id: {}", id);
        try {
            return weatherQueriesRepository.findById(id);
        } catch (Exception e) {
            logger.error("Error fetching weather query with id {}: {}", id, e.getMessage());
            throw new RuntimeException("Failed to fetch weather query", e);
        }
    }

    // Read (all entries)
    public List<WeatherQueries> getAllWeatherQueries() {
        logger.info("Fetching all weather query entries");
        try {
            return weatherQueriesRepository.findAll();
        } catch (Exception e) {
            logger.error("Error fetching all weather query entries: {}", e.getMessage());
            throw new RuntimeException("Failed to fetch all weather query entries", e);
        }
    }

    // Read all entries by user id
    public List<WeatherQueries> getAllWeatherQueriesByUserId(String userId) {
        logger.info("Fetching all weather query entries by userId");
        try {
            return weatherQueriesRepository.findAllByUserId(userId);
        } catch (Exception e) {
            logger.error("Error fetching all weather query entries: {}", e.getMessage());
            throw new RuntimeException("Failed to fetch all weather query entries", e);
        }
    }

    // Update
    public WeatherQueries updateWeatherQuery(Long id, String query) {
        logger.info("Updating weather query with id: {}", id);
        try {
            Optional<WeatherQueries> existingWeatherQuery = weatherQueriesRepository.findById(id);
            if (existingWeatherQuery.isPresent()) {
                WeatherQueries weatherQueryToUpdate = existingWeatherQuery.get();
                // Update the fields you want to change
                weatherQueryToUpdate.setQuery(query);
                // Add more fields as necessary
                return weatherQueriesRepository.save(weatherQueryToUpdate);
            } else {
                logger.warn("Weather query with id {} not found", id);
                throw new RuntimeException("Weather query not found");
            }
        } catch (Exception e) {
            logger.error("Error updating weather query with id {}: {}", id, e.getMessage());
            throw new RuntimeException("Failed to update weather query", e);
        }
    }

    // Delete
    public void deleteWeatherQuery(Long id) {
        logger.info("Deleting weather query with id: {}", id);
        try {
            if (weatherQueriesRepository.existsById(id)) {
                weatherQueriesRepository.deleteById(id);
            } else {
                logger.warn("Weather query with id {} not found", id);
                throw new RuntimeException("Weather query not found");
            }
        } catch (Exception e) {
            logger.error("Error deleting weather query with id {}: {}", id, e.getMessage());
            throw new RuntimeException("Failed to delete weather query", e);
        }
    }

}

File: backend-kotlin/src/main/kotlin/com/nbweatherproject/backend/services/WeatherQueriesService.kt
package com.nbweatherproject.backend

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.Optional

@Service
class WeatherQueriesService(private val weatherQueriesRepository: WeatherQueriesRepository) {

    companion object {
        private val logger = LoggerFactory.getLogger(WeatherQueriesService::class.java)
    }

    // Create
    fun createWeatherQuery(query: String, userId: String): WeatherQueries {
        logger.info("Creating or retrieving weather query entry")
        return try {
            val existingQuery = weatherQueriesRepository
                .findByQueryAndUserId(query, userId)

            if (existingQuery.isPresent) {
                logger.info("Existing weather query found for query: {} and userId: {}",
                    query, userId)
                existingQuery.get()
            } else {
                logger.info("Creating new weather query entry for query: {} and userId: {}",
                    query, userId)
                weatherQueriesRepository.save(query, userId)
            }
        } catch (e: Exception) {
            logger.error("Error creating or retrieving weather query entry: {}", e.message)
            throw RuntimeException("Failed to create or retrieve weather query entry", e)
        }
    }

    // Read (single entry)
    fun getWeatherQueryById(id: Long): Optional<WeatherQueries> {
        logger.info("Fetching weather query with id: {}", id)
        return try {
            weatherQueriesRepository.findById(id)
        } catch (e: Exception) {
            logger.error("Error fetching weather query with id {}: {}", id, e.message)
            throw RuntimeException("Failed to fetch weather query", e)
        }
    }

    // Read (all entries)
    fun getAllWeatherQueries(): List<WeatherQueries> {
        logger.info("Fetching all weather query entries")
        return try {
            weatherQueriesRepository.findAll()
        } catch (e: Exception) {
            logger.error("Error fetching all weather query entries: {}", e.message)
            throw RuntimeException("Failed to fetch all weather query entries", e)
        }
    }

    // Read all entries by user id
    fun getAllWeatherQueriesByUserId(userId: String): List<WeatherQueries> {
        logger.info("Fetching all weather query entries by userId")
        return try {
            weatherQueriesRepository.findAllByUserId(userId)
        } catch (e: Exception) {
            logger.error("Error fetching all weather query entries: {}", e.message)
            throw RuntimeException("Failed to fetch all weather query entries", e)
        }
    }

    // Update
    fun updateWeatherQuery(id: Long, query: String): WeatherQueries {
        logger.info("Updating weather query with id: {}", id)
        return try {
            val existingWeatherQuery = weatherQueriesRepository.findById(id)
            if (existingWeatherQuery.isPresent) {
                val weatherQueryToUpdate = existingWeatherQuery.get()
                // Update the fields you want to change
                weatherQueryToUpdate.query = query
                // Add more fields as necessary
                weatherQueriesRepository.save(weatherQueryToUpdate)
            } else {
                logger.warn("Weather query with id {} not found", id)
                throw RuntimeException("Weather query not found")
            }
        } catch (e: Exception) {
            logger.error("Error updating weather query with id {}: {}", id, e.message)
            throw RuntimeException("Failed to update weather query", e)
        }
    }

    // Delete
    fun deleteWeatherQuery(id: Long) {
        logger.info("Deleting weather query with id: {}", id)
        try {
            if (weatherQueriesRepository.existsById(id)) {
                weatherQueriesRepository.deleteById(id)
            } else {
                logger.warn("Weather query with id {} not found", id)
                throw RuntimeException("Weather query not found")
            }
        } catch (e: Exception) {
            logger.error("Error deleting weather query with id {}: {}", id, e.message)
            throw RuntimeException("Failed to delete weather query", e)
        }
    }

    // Search
    fun searchWeatherQueries(query: String): List<WeatherQueries> {
        return weatherQueriesRepository.findByQueryContainingIgnoreCase(query)
    }
}

